More distributions

dirmeier opened the pull request:
Hello all,
this PR implements Weibull, BetaProportion, NegativeBinomial, NegativeBinomial2 and ZeroInflatedNegativeBinomial distributions. The BetaProportion is a reparameterization using a mean and a sample size parameter (https://en.wikipedia.org/wiki/Beta_distribution#Mean_and_sample_size) and can be used for regression. The NB2 is an alternative parameterization using mean and dispersion parameters and also useful for regression.

All of these are frequently used in Bayesian stats and it would be nice if you could merge them. Thanks!

Best regards,
Simon

fehiepsi reviewed (commented):
Awesome work, @dirmeier! Thanks for your contribution. Overall looks great to me. I just have some opinions on the name of some parameters and some simplifications.

fehiepsi commented on numpyro/distributions/conjugate.py:
```diff
@@ -209,3 +214,72 @@ def mean(self):
+            - jnp.log(post_value)
+            + self.concentration * (-self.logits)
+            - post_value * jnp.log1p(nn.softplus(-self.logits))
+        )
```
I think that you can use this formula
```
return -(self.total_count * softplus(self.logits) + value * softplus(-self.logits) + _log_beta_1(self.concentration, value))
```

dirmeier commented:
Ciao @fehiepsi, I think I addressed all of your comments. I am still receiving some errors on the unit tests, such as:

```
FAILED test/test_distributions.py::test_jit_log_likelihood[NegativeBinomialLogits-None-params116] - AssertionError: 
FAILED test/test_distributions.py::test_jit_log_likelihood[NegativeBinomialLogits-None-params118] - AssertionError: 
FAILED test/test_distributions.py::test_jit_log_likelihood[NegativeBinomial2-None-params120] - AssertionError: 
FAILED test/test_distributions.py::test_gof[BetaProportion-<lambda>-params4] - assert 0.0 > 2e-05
FAILED test/test_distributions.py::test_gof[BetaProportion-<lambda>-params5] - assert 0.0 > 2e-05
```

Could you give me a hint what the issues are with these? Thanks!

Best,
Simon

dirmeier commented on numpyro/distributions/discrete.py:
```diff
@@ -826,3 +827,125 @@ def Geometric(probs=None, logits=None, validate_args=None):
+    arg_constraints = {"alpha": constraints.positive, "beta": constraints.positive}
+    support = constraints.nonnegative_integer
+
+    def __init__(self, alpha, beta, validate_args=None):
```
I needed to move these into `conjugate.py` in order to avoid circular imports. Is that ok?

fehiepsi replied:
Yup, I imagined circular import happens, so I used
```
import numpyro.distributions.conjugate as conjugate
```
but it is good to know that solution doesn't work. Thanks for clarifying!

fehiepsi commented on numpyro/distributions/__init__.py:
```diff
@@ -5,10 +5,15 @@
     GammaPoisson,
+    NegativeBinomial2,
+    NegativeBinomialLogits,
+    NegativeBinomialProbs,
```
It would be nice to have `NegativeBinomial` which dispatches to either of those two classes. Similar to `Bernoulli` dispatching to `BernoulliLogits` and `BernoulliProbs`

dirmeier commented on numpyro/distributions/continuous.py:
```diff
@@ -1426,3 +1426,88 @@ def infer_shapes(low=(), high=()):
+    support = constraints.real
+    reparametrized_params = ["scale", "concentration"]
+
+    def __init__(self, scale=0.0, concentration=1.0, validate_args=None):
```
Ah, nice catch. No, I think it's better without. Bringing people to actively think about parameterization is better than default arguments. But your call.

dirmeier commented on numpyro/distributions/discrete.py:
```diff
@@ -826,3 +827,125 @@ def Geometric(probs=None, logits=None, validate_args=None):
+        return bt
+
+
+class ZeroInflatedNegativeBinomial(Distribution):
```
Cool! Thanks for the hint.

fehiepsi reviewed (commented):
Looks great (and clean) to me overall. Could you help me add new entries to [docs](https://github.com/pyro-ppl/numpyro/blob/master/docs/source/distributions.rst)? I'll look at the failing tests.

fehiepsi commented:
For failing tests, I guess you can do
+ For `test_jit_log_likelihood`, change `atol` from 1e-5 to 2e-5.
+ For gof test, I guess we can increase the value of `concentration`? It seems to me that 10000 samples are not enough for a small `concentration`.

fehiepsi reviewed (dismissed):
LGTM pending new entries in the [docs file](https://github.com/pyro-ppl/numpyro/blob/master/docs/source/distributions.rst).

dirmeier commented:
Ciao @fehiepsi. Yes thanks,  just added them.  The rest is also done I believe. An thanks for the quick reviews.  It's a please to contribute to a project, when the devs are so active :)

fehiepsi commented:
Thanks, @dirmeier!! My impression through your implementations is you understand numpyro pretty well. I do love to see more contributions from you in the future. :)
